pennylane/interfaces/tensorflow.py

```diff
@@ -116,207 +165,110 @@ def _res_restructured(res, tapes):
+    res = _to_tensors(execute_fn(numpy_tapes))
 
     @tf.custom_gradient
-    def _execute(*parameters):  # pylint:disable=unused-argument
-        if logger.isEnabledFor(logging.DEBUG):
-            logger.debug(
-                "Entry with args=(parameters=%s) called by=%s",
-                parameters,
-                "::L".join(
-                    str(i) for i in inspect.getouterframes(inspect.currentframe(), 2)[1][1:3]
-                ),
-            )
-
-        def grad_fn(*dy, **tfkwargs):
-            """Returns the vector-Jacobian product with given
-            parameter values and output gradient dy"""
-            if logger.isEnabledFor(logging.DEBUG):
-                logger.debug(
-                    "Entry with args=(dy=%s, tfkwargs=%s) called by=%s",
-                    dy,
-                    tfkwargs,
-                    "::L".join(
-                        str(i) for i in inspect.getouterframes(inspect.currentframe(), 2)[1][1:3]
-                    ),
-                )
+    def custom_gradient_execute(*parameters):  # pylint:disable=unused-argument
+        """An execution of tapes with VJP's registered with tensorflow.
 
-            # whether the tapes contain multiple measurements
-            multi_measurements = [len(tape.measurements) > 1 for tape in tapes]
+        Closure:
+            tapes (tuple(QuantumTape))
+            numpy_tapes (tuple(QuantumTape)): tapes but with numpy parameters
+            numpy_params (list(numpy.ndarray)): numpy versions of ``parameters``.
+            jpc (JacobianProductCalculator): a class that can calculate the VJP.
 
+        """
+
+        def vjp_fn(*dy, **tfkwargs):
             # TF obeys the dL/dz_conj convention instead of the
             # dL/dz convention of PennyLane, autograd and jax. This converts between the formats
             dy = _recursive_conj(dy)
 
-            # reconstruct the nested structure of dy
-            dy = _res_restructured(dy, tapes)
+            if not differentiable:
+                inner_tapes = numpy_tapes
+            elif not context.executing_eagerly():
+                warnings.warn(
+                    "PennyLane does not provide the higher order derivatives of tensorflow jacobians."
+                )
+                # Using numpy_tapes instead seems to cause failures
+                inner_tapes = set_parameters_on_copy(tapes, numpy_params)
+            else:
+                inner_tapes = tapes
 
-            if jacs:
-                # Jacobians were computed on execution
-                # No additional quantum evaluations needed; simply compute the VJPs directly.
-                vjps = _compute_vjp(dy, jacs, multi_measurements, has_partitioned_shots)
+            # reconstruct the nested structure of dy
+            nested_dy = _res_restructured(dy, tapes)
+
+            try:
+                vjps = jpc.compute_vjp(inner_tapes, nested_dy)
+            except AttributeError as e:

```

Comment:
While not ideal, the alternative of using a `tf.py_function` causes an extra VJP call to occur. While caching could eliminate the biggest performance issues from having an extra call to `vjp_fn`, caching might be turned off. 

But it might an option to consider in the future.

Reply:
let's just add a story to track the decision in the future. side note: is the try-except the only reliable way to know when this error will happen ahead of time?

Reply:
We could check if we are in eager execution and using a device vjp.

But then we start caring about the way we are calculating the vjp here, which this switchover was supposed to get rid of.